tfhe/src/core_crypto/commons/math/decomposition/decomposer.rs

```diff
@@ -219,15 +227,26 @@ where
         level_count: DecompositionLevelCount,
         ciphertext_modulus: CiphertextModulus<Scalar>,
     ) -> Self {
-        debug_assert!(
-            Scalar::BITS > base_log.0 * level_count.0,
-            "Decomposed bits exceeds the size of the integer to be decomposed"
+        assert!(
+            !ciphertext_modulus.is_power_of_two(),
+            "Got a power of 2 modulus as input for SignedDecomposerNonNative, \
+            this is not supported, use SignedDecomposer instead"
         );
-        Self {
+
+        let sself = Self {
             base_log: base_log.0,
             level_count: level_count.0,
             ciphertext_modulus,
-        }
+        };
+
+        let log2_containing_modulus = sself.ciphertext_modulus_bit_count();
+
+        debug_assert!(
+            log2_containing_modulus > (base_log.0 * level_count.0) as u32,
+            "Decomposed bits exceeds the size of the integer modulus"
+        );
```

Comment:
c.f. comment in tests.rs file regarding this.

Reply:
so I think because we do the rounding before it does not make sense to have log2_containing_modulus == base_log * level_count

e.g. on a small example

beta = 2,
level = 2

```
b7 | b6 | b5 | b4 | b3 | b2 | b1 | b0
                    ^
                    |
                    rounding bit
```

if we have beta * level = 8 then the rounding won't work with our current implementation

On a different note it would mean that we are decomposing all the bits of the input :) so not sure it's a problem in practice to have one less bit

Reply:
Yes, this means we get an exact decomposition of the input which is unlikely to ever be used in practice. But theoretically its valid to perform an exact decomposition in this way and the rounding (i.e. closest_representable) would be the same as the input. If you check the code for closest_representable you would get `shift_to_native` being 0 and then we would input this in native_closest_representable. Here `shift` would be -1 and I don't know what input >> -1 is in practice. Intuitively it should be the same as input << 1. If the final res << -1 is also just res >> 1then everything works as expected and the output is the same as the input.

Reply:
If it's too much of a headache you can leave the inequality strict as I don't think we would use such parameters but on the other hand we don't want a restriction where one isn't necessary.

Reply:
agreed, I don't think the shift behaves well with negative values though

We might keep it as is for now, and if the need arises we'll refactor the code